sdks/python/apache_beam/transforms/util.py

```diff
@@ -161,3 +173,196 @@ def RemoveDuplicates(pcoll):  # pylint: disable=invalid-name
+
+    # Avoid nonsensical or division-by-zero errors below due to noise.
+    a = max(a, 1e-10)
+    b = max(b, 1e-20)
+
+    last_batch_size = self._data[-1][0]
+    cap = min(last_batch_size * self.MAX_GROWTH_FACTOR, self._max_batch_size)
+
+    if self._target_batch_duration:
+      cap = min(cap, (self._target_batch_duration - a) / b)
+
+    if self._target_batch_overhead:
+      cap = min(cap, a / b * (1 / self._target_batch_overhead - 1))
+
+    # Avoid getting stuck at min_batch_size.
+    jitter = len(self._data) % 2
+    return int(max(self._min_batch_size + jitter, cap))
+
+
+class _GlobalWindowsBatchingDoFn(DoFn):
+  def __init__(self, batch_size_estimator):
+    self._batch_size_estimator = batch_size_estimator
+
+  def start_bundle(self):
+    self._batch = []
+    self._batch_size = self._batch_size_estimator.next_batch_size()
+    self._first = True
+
+  def process(self, element):
+    self._batch.append(element)
+    if len(self._batch) >= self._batch_size:
+      if self._first:
+        # This often involves non-trivial setup.
+        yield self._batch
+        self._first = False
+      else:
+        with self._batch_size_estimator.record_time(self._batch_size):
+          yield self._batch
+      self._batch = []
+      self._batch_size = self._batch_size_estimator.next_batch_size()
+
+  def finish_bundle(self):
+    if self._batch:
+      with self._batch_size_estimator.record_time(self._batch_size):
+        yield window.GlobalWindows.windowed_value(self._batch)
+      self._batch = None
+      self._batch_size = self._batch_size_estimator.next_batch_size()
+
+
+class _WindowAwareBatchingDoFn(DoFn):
+  def __init__(self, batch_size_estimator):
+    self._batch_size_estimator = batch_size_estimator
+
+  def start_bundle(self):
+    self._batches = collections.defaultdict(list)
+    self._batch_size = self._batch_size_estimator.next_batch_size()
+    self._first = True
+
+  def process(self, element, window=DoFn.WindowParam):
+    self._batches[window].append(element)
+    if len(self._batches[window]) >= self._batch_size:
+      windowed_batch = windowed_value.WindowedValue(
+          self._batches[window], window.max_timestamp(), (window,))
+      if self._first:
+        yield windowed_batch
+        self._first = False
+      else:
+        with self._batch_size_estimator.record_time(self._batch_size):
+          yield windowed_batch
+      del self._batches[window]
+      self._batch_size = self._batch_size_estimator.next_batch_size()
+
+  def finish_bundle(self):
+    for window, batch in self._batches.items():
+      if batch:
+        with self._batch_size_estimator.record_time(self._batch_size):
+          yield windowed_value.WindowedValue(
+              batch, window.max_timestamp(), (window,))
+    self._batches = None
+    self._batch_size = self._batch_size_estimator.next_batch_size()
+
+
+@typehints.with_input_types(T)
+@typehints.with_output_types(typehints.List[T])
+class BatchElements(PTransform):
```

Comment:
Is there a way to also make sure that the serialized form of a batch doesn't get too big?
Specifically to go around cPickle limitations

Reply:
It's not possible to access Dataflow's PCollection sizing here, so I'd have to replicate the logic. We'd also expect such elements to be "large" in time as well. 

Measuring by time is better than size because it allows us to avoid batching beyond the point where it does any good (the overhead limit, which is also nice because it's unitless and so has a reasonable default) and to a point where it hurts liquid sharding and progress (the duration limit).

Reply:
Great point, sounds good.